hack/generator/pkg/codegen/pipeline_augment_status.go

```diff
@@ -0,0 +1,421 @@
+	statusVisitor := makeStatusVisitor()
+
+	var otherTypes []astmodel.TypeDefinition
+	for typeName, typeDef := range swaggerTypes.types {
+		newName := appendStatusToName(typeName)
+		otherTypes = append(otherTypes, astmodel.MakeTypeDefinition(newName, statusVisitor.Visit(typeDef.Type(), nil)))
+	}
+
+	resources := make(resourceLookup)
+	for resourceName, resourceDef := range swaggerTypes.resources {
+		resources[resourceName] = statusVisitor.Visit(resourceDef.Type(), nil)
+	}
+
+	return resources, otherTypes
+}
+
+func appendStatusToName(typeName astmodel.TypeName) astmodel.TypeName {
+	return astmodel.MakeTypeName(typeName.PackageReference, typeName.Name()+"_Status")
+}
+
+func makeStatusVisitor() astmodel.TypeVisitor {
+	visitor := astmodel.MakeTypeVisitor()
+
+	visitor.VisitTypeName = func(this *astmodel.TypeVisitor, it astmodel.TypeName, ctx interface{}) astmodel.Type {
+		return appendStatusToName(it)
+		/*
+			if statusType, ok := statusTypes[it]; ok {
+				switch statusType.Type().(type) {
+				case *astmodel.ResourceType:
+				case *astmodel.ObjectType:
+					return appendStatusToName(it)
+				}
+			}
+
+			return it
+		*/
+	}
+
+	return visitor
+}
+
+var swaggerVersionRegex = regexp.MustCompile("\\d{4}-\\d{2}-\\d{2}(-preview)?")
+
+type swaggerTypes struct {
+	resources astmodel.Types
+	types     astmodel.Types
+}
+
+func loadSwaggerData(ctx context.Context, idFactory astmodel.IdentifierFactory, config *config.Configuration) (swaggerTypes, error) {
+
+	result := swaggerTypes{
+		resources: make(astmodel.Types),
+		types:     make(astmodel.Types),
+	}
+
+	cache := jsonast.MakeOpenAPISchemaCache()
+
+	for _, namespace := range config.Status.Schemas {
+		rootPath := path.Join(config.Status.SchemaRoot, namespace.BasePath)
+
+		// based off: https://github.com/Azure/azure-resource-manager-schemas/blob/1153d20f38e11bf1815bb855f309f83023fe6b0c/generator/cmd/listbasepaths.ts#L12-L15
+		if _, err := os.Stat(path.Join(rootPath, "readme.md")); os.IsNotExist(err) {
+			return swaggerTypes{}, errors.Errorf("configuration points to a directory that does not have status: %q", rootPath)
+		}
+
+		versionDirs, err := findVersionDirectories(rootPath)
+		if err != nil {
+			return swaggerTypes{}, errors.Wrapf(err, "unable to find version directories")
+		}
+
+		for _, versionDir := range versionDirs {
+			version := path.Base(versionDir)
+
+			schemas, err := loadSchemas(ctx, versionDir, cache)
+			if err != nil {
+				return swaggerTypes{}, err
+			}
+
+			outputGroup := namespace.Namespace
+			if namespace.Suffix != "" {
+				outputGroup = outputGroup + "." + namespace.Suffix
+			}
+
+			extractor := typeExtractor{
+				outputVersion: version,
+				outputGroup:   outputGroup,
+				idFactory:     idFactory,
+				cache:         cache,
+				config:        config,
+			}
+
+			for filePath, schema := range schemas {
+				err := extractor.extractTypes(ctx, filePath, schema, result.resources, result.types)
```

Comment:
I'm sometimes thrown by functions like this where the parameter is also being modified by the function and contains the output/result.

I'm not sure there's really a good alternative though for maps (for arrays you could return the stuff to append and then append it on the next line with I usually prefer)

Reply:
Also unlike C/C++ (where I see this pattern quite frequently), go doesn't have a way to annotate the function signature/parameters with `const` (or in this case, _not_) to inform callers that actually that parameter will change.

Reply:
Going to leave as-is.